### PR TITLE
First steps toward quadrature

### PR DESCRIPTION
**Pull request by nshaffer**
Here is a starting point for the proposed quadrature module (#112)

- `trapz`, `trapz_weights`
  - Simple API, implementation, and tests for 1-d real arrays only
  - TODO: complex arrays (straightforward for 1-d)
  - TODO: multi-dimensional arrays (still some API questions to resolve)

- `simps`, `simps_weights`
  - API for 1-d real arrays
  - TODO: Implementation & tests
  - TODO: complex arrays
  - TODO: multi-dimensional arrays

I did not get as far as I hoped in the past week, put I want to get some baseline code up for others to work off of. 

Some points I'm hoping for some specific feedback on:
- Edge cases. Do we like the proposed handling of small arrays?
- The `even` argument of `simps` and `simps_weights`. Does this seem like a reasonable thing to do? Does the API make sense?

### CODE REVIEW

**Comment by nshaffer** [comment]
@certik Hm, I'm of two minds about that. Sometimes you're testing that a function return a specific value, and it is incorrect behavior if it differs even by one bit. That's the case in `optval` and for my edge-case tests for `trapz`. But I do like the ideal of warning-free compilation. I've gone ahead and changed all the real equality tests here to tolerance checks against `epsilon`.

File: src/stdlib_experimental_quadrature_trapz.fypp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        real(${KIND}$), dimension(size(x)) :: w
+
+        integer :: i
+        integer :: n
+
+        n = size(x)
+
+        select case (n)
+        case (0)
+            ! no action needed
```
Comment (jvdp1) [inline src/stdlib_experimental_quadrature_trapz.fypp]: I agree that "no action" is needed. But should the function not return somethin for `w` (e.g. NaN, a huge number, 0,...)? Currently `w` is undefined.

The other functions return 0 for `case (0)`.

Comment (nshaffer) [reply src/stdlib_experimental_quadrature_trapz.fypp]: For `trapz`, zero-size inputs behave the same as summing a zero-size array, producing zero. For `trapz_weights`, you want the result to behave such that `sum(w*x)` also produces zero. 

There are also practical challenges to returning anything besides a zero-size array. It's OK that it's value is undefined. That's just how zero-size arrays go.
